File: apps/web/src/api/interactives.ts
import { postGraphQL } from "./graphqlClient";
import { createToken } from "./tokens";
import type { TokenRecord } from "@/types/token";

/**
 * Authoring, activating and approving interactive elements.
 *
 * Spec 030. Every rule this module appears to describe is actually enforced on
 * the server — a player is refused a locked door there, not here — so nothing
 * below checks permission. A second opinion in the client would be a second
 * thing to keep right, and the one that drifts is always the one people
 * believe.
 *
 * What this module does own is the *shape*: the authoring form is built from
 * `effectRegistry` rather than from a list written here, so a Game Master is
 * offered exactly what this build can perform and never an option that would
 * silently do nothing (FR-038).
 */

/** Which subjects an effect may attach to. */
export type SubjectKind = "prop" | "door" | "region";

/** How a configuration field is filled in. */
export type ConfigFieldKind =
  | "boolean"
  | "choice"
  | "reference"
  | "referenceList";

export interface ChoiceOption {
  value: string;
  label: string;
}

export interface ConfigField {
  key: string;
  label: string;
  kind: ConfigFieldKind;
  /** What a reference points at — `wall`, `light`, `loreEntry`, `scene`. */
  referenceOf: string | null;
  options: ChoiceOption[] | null;
  required: boolean;
}

/**
 * One effect this build can perform.
 *
 * Note what is absent: any notion of an effect being unavailable. The registry
 * is the union of what is compiled in, so a subsystem that does not exist
 * contributes nothing and there is no dead option to grey out.
 */
export interface EffectDeclaration {
  id: string;
  label: string;
  description: string;
  subjectKinds: SubjectKind[];
  config: ConfigField[];
}

/**
 * One authored interactive.
 *
 * A player receives a narrower version of this: the Game-Master-only fields
 * arrive as `null`. That is an interface boundary rather than a security one —
 * per the spec, secrets are a table concern — but sending a player an effect's
 * configuration would invite some future client to render it.
 */
export interface Interactive {
  interactiveId: string;
  sceneId: string;
  subjectKind: SubjectKind;
  subjectRef: string | null;
  geometry: unknown | null;
  trigger: "click" | "enter";
  effectId: string | null;
  effectConfig: Record<string, unknown> | null;
  activation: string | null;
  fireMode: string | null;
  firedAt: string | null;
  /** Whether the subsystem that performs the effect is in this build. */
  available: boolean | null;
  /** Whether this viewer's activation would do anything. A hint, not a right. */
  canActivate: boolean;
}

/** What happened, as a tagged outcome rather than a boolean. */
export interface ActivationResult {
  outcome: "performed" | "requested" | "refused" | "unavailable" | "noEffect";
  /** `gmOnly`, `locked` or `alreadyFired`. Present only when refused. */
  reason: string | null;
  requestId: string | null;
  effectId: string | null;
  effectConfig: Record<string, unknown> | null;
  /**
   * Things the Game Master should know about what just ran.
   *
   * Always empty for a player. These are notes about the *authoring* — a
   * switch naming a lamp that has been deleted — and a player has no use for
   * one and no way to act on it.
   */
  notices: string[];
}

const INTERACTIVE_FIELDS = `
  interactiveId
  sceneId
  subjectKind
  subjectRef
  geometry
  trigger
  effectId
  effectConfig
  activation
  fireMode
  firedAt
  available
  canActivate
`;

export async function getEffectRegistry(): Promise<EffectDeclaration[]> {
  const payload = await postGraphQL<{ effectRegistry: EffectDeclaration[] }>(
    `query {
      effectRegistry {
        id
        label
        description
        subjectKinds
        config { key label kind referenceOf options { value label } required }
      }
    }`,
    {},
  );
  return payload?.effectRegistry ?? [];
}

export async function getInteractives(sceneId: string): Promise<Interactive[]> {
  const payload = await postGraphQL<{ interactives: Interactive[] }>(
    `query ($sceneId: UUID!) {
      interactives(sceneId: $sceneId) { ${INTERACTIVE_FIELDS} }
    }`,
    { sceneId },
  );
  return payload?.interactives ?? [];
}

export interface CreateInteractiveInput {
  sceneId: string;
  subjectKind: SubjectKind;
  subjectRef?: string | null;
  geometry?: unknown;
  effectId?: string | null;
  effectConfig?: Record<string, unknown> | null;
  trigger: "click" | "enter";
  activation: string;
  fireMode?: string | null;
}

export async function createInteractive(
  input: CreateInteractiveInput,
): Promise<Interactive> {
  const payload = await postGraphQL<{ createInteractive: Interactive }>(
    `mutation ($input: GraphQLCreateInteractiveInput!) {
      createInteractive(input: $input) { ${INTERACTIVE_FIELDS} }
    }`,
    { input },
  );
  return payload.createInteractive;
}

export interface UpdateInteractiveInput {
  geometry?: unknown;
  effectId?: string | null;
  effectConfig?: Record<string, unknown> | null;
  trigger?: string | null;
  activation?: string | null;
  fireMode?: string | null;
  /**
   * Turn an interactive back into scenery.
   *
   * Needed because an absent `effectId` means "leave it alone" in a partial
   * update, so there would otherwise be no way to say "remove it".
   */
  clearEffect?: boolean;
}

export async function updateInteractive(
  interactiveId: string,
  input: UpdateInteractiveInput,
): Promise<Interactive> {
  const payload = await postGraphQL<{ updateInteractive: Interactive }>(
    `mutation ($id: UUID!, $input: GraphQLUpdateInteractiveInput!) {
      updateInteractive(interactiveId: $id, input: $input) { ${INTERACTIVE_FIELDS} }
    }`,
    { id: interactiveId, input },
  );
  return payload.updateInteractive;
}

export async function deleteInteractive(
  interactiveId: string,
): Promise<boolean> {
  const payload = await postGraphQL<{ deleteInteractive: boolean }>(
    `mutation ($id: UUID!) { deleteInteractive(interactiveId: $id) }`,
    { id: interactiveId },
  );
  return payload?.deleteInteractive ?? false;
}

export async function resetInteractive(
  interactiveId: string,
): Promise<Interactive> {
  const payload = await postGraphQL<{ resetInteractive: Interactive }>(
    `mutation ($id: UUID!) {
      resetInteractive(interactiveId: $id) { ${INTERACTIVE_FIELDS} }
    }`,
    { id: interactiveId },
  );
  return payload.resetInteractive;
}

/**
 * The one mutation a player calls.
 *
 * Returns what happened rather than whether it worked, because "it did not
 * run" covers four different situations and a player told only "no" cannot
 * tell a locked door from a broken product.
 */
export async function activateInteractive(
  interactiveId: string,
): Promise<ActivationResult> {
  const payload = await postGraphQL<{
    activateInteractive: ActivationResult;
  }>(
    `mutation ($id: UUID!) {
      activateInteractive(interactiveId: $id) {
        outcome
        reason
        requestId
        effectId
        effectConfig
        notices
      }
    }`,
    { id: interactiveId },
  );
  return payload.activateInteractive;
}

/**
 * Place a prop: a book, a chest, a lever.
 *
 * A prop is a token of the existing `object` kind with no actor. That is the
 * whole implementation, and deliberately so — token placement, artwork,
 * movement, ordering and live sync all exist and work, and a parallel "props"
 * concept would duplicate every one of them.
 *
 * The consequence to watch is that anything consuming tokens must tolerate one
 * with no actor. Spec 029 established that precedent: `tokenStatus` skips
 * actorless tokens, calling them markers rather than creatures.
 *
 * This wrapper exists to name the intent, not to add behaviour. If it ever
 * grows a second step, that is the signal a prop has stopped being a token.
 */
export function placeProp(
  sceneId: string,
  x: number,
  y: number,
  photoUrl?: string,
): Promise<TokenRecord> {
  return createToken({ sceneId, x, y, tokenType: "object", photoUrl });
}

/**
 * Make a wall a door, or stop it being one (FR-007).
 *
 * Designating also gives the door an interactive, so clicking it opens it
 * without the Game Master having to author one — a door nobody can touch is
 * not what "designate a door" means to anybody.
 */
export async function setDoorDesignation(
  wallId: string,
  isDoor: boolean,
): Promise<boolean> {
  const payload = await postGraphQL<{ setDoorDesignation: boolean }>(
    `mutation ($wallId: UUID!, $isDoor: Boolean!) {
      setDoorDesignation(wallId: $wallId, isDoor: $isDoor)
    }`,
    { wallId, isDoor },
  );
  return payload?.setDoorDesignation ?? false;
}

/** Change who may open a door, not whether it is open (FR-013). */
export async function setDoorLock(
  wallId: string,
  locked: boolean,
): Promise<boolean> {
  const payload = await postGraphQL<{ setDoorLock: boolean }>(
    `mutation ($wallId: UUID!, $locked: Boolean!) {
      setDoorLock(wallId: $wallId, locked: $locked)
    }`,
    { wallId, locked },
  );
  return payload?.setDoorLock ?? false;
}

/** Hide a door from the table, or show it. */
export async function setDoorSecret(
  wallId: string,
  secret: boolean,
): Promise<boolean> {
  const payload = await postGraphQL<{ setDoorSecret: boolean }>(
    `mutation ($wallId: UUID!, $secret: Boolean!) {
      setDoorSecret(wallId: $wallId, secret: $secret)
    }`,
    { wallId, secret },
  );
  return payload?.setDoorSecret ?? false;
}

/**
 * What to tell somebody whose activation did not run.
 *
 * `null` for an outcome that needs nothing said. Silence for a *refusal* is
 * indistinguishable from the product being broken, which is the whole reason
 * the outcome is tagged rather than boolean (FR-014) — but silence for a
 * performed effect is correct, because the effect is the feedback.
 */
export function refusalNotice(result: ActivationResult): string | null {
  switch (result.outcome) {
    case "performed":
      return null;
    case "requested":
      return "Asked the GM. Nothing happens until they say so.";
    case "unavailable":
      // Not a failure the player caused, and not one they can do anything
      // about. Saying so beats a click that appears to do nothing.
      return "This does not work in this session.";
    case "noEffect":
      return null;
    case "refused":
      switch (result.reason) {
        case "locked":
          return "It is locked.";
        case "gmOnly":
          return "Only the GM can do this.";
        case "alreadyFired":
          return "This has already happened.";
        default:
          return "That did not work.";
      }
    default:
      return null;
  }
}

File: src/server/src/graphql/queries/interactives.rs
//! `effectRegistry` and `interactives(sceneId)` — what a build can perform,
//! and what is authored on a scene.
//!
//! # Two different answers to one query
//!
//! A Game Master receives the authoring view: subject, effect, configuration,
//! fire state, and whether the effect is still available. A player receives
//! which subjects are interactive and whether they may activate them — not the
//! effect, not its configuration, not what it targets.
//!
//! That split is not a security boundary. Per the spec, secrets are a table
//! concern and the geometry reaches every client anyway. It is an *interface*
//! boundary: a player has no use for an effect's configuration, and sending it
//! would invite some future client to render it.

use async_graphql::{Context, Error, Json, Object, Result as GraphQLResult, SimpleObject};
use diesel::prelude::*;
use uuid::Uuid;

use thunderforge_canvas_core::interaction::{ActivationOutcome, ConfigFieldKind};

use crate::graphql::{app_state, authenticated_user};

#[derive(SimpleObject, Debug, Clone)]
pub struct GraphQLChoiceOption {
    pub value: String,
    pub label: String,
}

/// One field the authoring form renders.
///
/// Flattened rather than expressed as a GraphQL union because the shapes
/// differ only in which single extra field they carry, and a four-member union
/// would cost every client an inline fragment to read one string.
#[derive(SimpleObject, Debug, Clone)]
pub struct GraphQLConfigField {
    pub key: String,
    pub label: String,
    /// `boolean`, `choice`, `reference` or `referenceList`.
    pub kind: String,
    /// What a reference points at — `wall`, `light`, `loreEntry`, `scene`.
    /// Present for `reference` and `referenceList`.
    pub reference_of: Option<String>,
    /// Present for `choice`.
    pub options: Option<Vec<GraphQLChoiceOption>>,
    pub required: bool,
}

#[derive(SimpleObject, Debug, Clone)]
pub struct GraphQLEffectDeclaration {
    pub id: String,
    pub label: String,
    pub description: String,
    /// `prop`, `door`, `region`.
    pub subject_kinds: Vec<String>,
    pub config: Vec<GraphQLConfigField>,
}

/// One authored interactive, reduced to what this viewer has a use for.
#[derive(SimpleObject, Debug, Clone)]
pub struct GraphQLInteractive {
    pub interactive_id: Uuid,
    pub scene_id: Uuid,
    pub subject_kind: String,
    pub subject_ref: Option<Uuid>,
    /// The region's area. Game Master only — a region is not an annotation and
    /// players are never shown one.
    pub geometry: Option<Json<serde_json::Value>>,
    pub trigger: String,
    /// Game Master only.
    pub effect_id: Option<String>,
    /// Game Master only.
    pub effect_config: Option<Json<serde_json::Value>>,
    /// Game Master only.
    pub activation: Option<String>,
    /// Game Master only.
    pub fire_mode: Option<String>,
    /// Game Master only.
    pub fired_at: Option<chrono::NaiveDateTime>,
    /// Whether this build still has the subsystem that performs the effect.
    ///
    /// Game Master only, and shown as a state rather than repaired (FR-041).
    /// A player is told nothing, because to them it is simply not interactive.
    pub available: Option<bool>,
    /// Whether *this* viewer's activation would do anything.
    ///
    /// The client uses it to decide whether to offer a cursor. It is a hint,
    /// not a permission: the server refuses on its own account when the
    /// mutation is called, because a client that draws the button anyway must
    /// still be told no.
    pub can_activate: bool,
}

/// One request awaiting a Game Master's decision.
#[derive(SimpleObject, Debug, Clone)]
pub struct GraphQLInteractionRequest {
    pub request_id: Uuid,
    pub interactive_id: Uuid,
    pub scene_id: Uuid,
    pub requested_by: Uuid,
    /// The player's display name, so the queue reads as people rather than ids.
    pub requested_by_name: Option<String>,
    /// What would run if it were approved, in the GM's language.
    pub proposed: Option<String>,
    pub created_at: chrono::NaiveDateTime,
}

#[derive(Default)]
pub struct InteractiveQuery;

#[Object]
impl InteractiveQuery {
    /// What this build can perform.
    ///
    /// Drives the authoring form, so a Game Master is offered exactly what
    /// exists and never an option that would silently do nothing (FR-038).
    async fn effect_registry(
        &self,
        ctx: &Context<'_>,
    ) -> GraphQLResult<Vec<GraphQLEffectDeclaration>> {
        // Authenticated, and no more. Which effects a build compiled in is not
        // a secret about anybody's world.
        let _ = authenticated_user(ctx)?;
        Ok(crate::interaction::registry()
            .all()
            .map(declaration_to_graphql)
            .collect())
    }

    /// What is waiting on a decision in this scene. Game Master only.
    ///
    /// A player is not shown the queue. Their own outcome reaches them
    /// directly, and the rest of it is a list of what other people asked for —
    /// which is the Game Master's business and, at some tables, information
    /// the GM is deliberately not sharing yet.
    async fn pending_interaction_requests(
        &self,
        ctx: &Context<'_>,
        scene_id: Uuid,
    ) -> GraphQLResult<Vec<GraphQLInteractionRequest>> {
        let state = app_state(ctx)?;
        let auth_user = authenticated_user(ctx)?;
        let user_id = auth_user.user_id;
        let is_admin = auth_user.is_admin;
        let mut conn = state
            .db_pool
            .get()
            .map_err(|_| Error::new("Failed to get DB connection"))?;

        tokio::task::spawn_blocking(move || {
            use crate::schema::{interactives, users};

            if !crate::auth::world_membership::is_dm_of_scene(
                &mut conn, user_id, is_admin, scene_id,
            )
            .map_err(|_| Error::new("Failed to check permission"))?
            {
                return Err(Error::new("Only the Game Master sees the queue"));
            }

            let rows = crate::interaction::pending_for_scene(&mut conn, scene_id)
                .map_err(|e| Error::new(format!("Failed to load requests: {e}")))?;

            let mut out = Vec::with_capacity(rows.len());
            for row in rows {
                let requested_by_name: Option<String> = users::table
                    .filter(users::id.eq(row.requested_by))
                    .select(users::username)
                    .first(&mut conn)
                    .optional()
                    .ok()
                    .flatten();

                // What the effect is, rather than which effect it is. A GM
                // deciding mid-session should not have to translate
                // `nav.request_scene` in their head.
                let proposed = interactives::table
                    .filter(interactives::interactive_id.eq(row.interactive_id))
                    .select(interactives::effect_id)
                    .first::<Option<String>>(&mut conn)
                    .optional()
                    .ok()
                    .flatten()
                    .flatten()
                    .and_then(|id| {
                        crate::interaction::registry()
                            .get(&id)
                            .map(|d| d.label.clone())
                    });

                out.push(GraphQLInteractionRequest {
                    request_id: row.request_id,
                    interactive_id: row.interactive_id,
                    scene_id: row.scene_id,
                    requested_by: row.requested_by,
                    requested_by_name,
                    proposed,
                    created_at: row.created_at,
                });
            }
            Ok(out)
        })
        .await
        .map_err(|_| Error::new("Failed to spawn blocking task"))?
    }

    /// Every interactive on a scene, as this viewer may know it.
    async fn interactives(
        &self,
        ctx: &Context<'_>,
        scene_id: Uuid,
    ) -> GraphQLResult<Vec<GraphQLInteractive>> {
        let state = app_state(ctx)?;
        let auth_user = authenticated_user(ctx)?;
        let user_id = auth_user.user_id;
        let is_admin = auth_user.is_admin;
        let mut conn = state
            .db_pool
            .get()
            .map_err(|_| Error::new("Failed to get DB connection"))?;

        tokio::task::spawn_blocking(move || {
            use crate::schema::{scenes, walls};

            // Membership before anything is read. What a scene contains is
            // world-scoped.
            let world_id: Uuid = scenes::table
                .filter(scenes::scene_id.eq(scene_id))
                .select(scenes::world_id)
                .first(&mut conn)
                .map_err(|_| Error::new("Scene not found"))?;

            let actor = crate::auth::world_membership::actor_in_world(
                &mut conn, user_id, is_admin, world_id,
            );
            if actor.role.is_none() && !actor.is_site_admin {
                return Err(Error::new("Not a member of this world"));
            }
            let runs_the_world = actor.runs_the_world();

            let rows = crate::interaction::for_scene(&mut conn, scene_id)
                .map_err(|e| Error::new(format!("Failed to load interactives: {e}")))?;

            // Every locked wall in the scene, in one read rather than per row.
            let locked_walls: std::collections::HashSet<Uuid> = walls::table
                .filter(walls::scene_id.eq(scene_id))
                .filter(walls::locked.eq(true))
                .select(walls::wall_id)
                .load::<Uuid>(&mut conn)
                .unwrap_or_default()
                .into_iter()
                .collect();

            let mut out = Vec::with_capacity(rows.len());
            for row in rows {
                let subject_locked = row
                    .subject_ref
                    .is_some_and(|id| row.subject_kind == "door" && locked_walls.contains(&id));
                let loaded = crate::interaction::LoadedInteractive {
                    row,
                    subject_locked,
                };
                let can_activate = matches!(
                    loaded.outcome(runs_the_world),
                    ActivationOutcome::Performed | ActivationOutcome::Requested
                );

                // A region is invisible to players, always. Nothing about it
                // is sent, not even that it exists.
                if !runs_the_world && loaded.row.subject_kind == "region" {
                    continue;
                }

                out.push(to_graphql(loaded, runs_the_world, can_activate));
            }
            Ok(out)
        })
        .await
        .map_err(|_| Error::new("Failed to spawn blocking task"))?
    }
}

fn to_graphql(
    loaded: crate::interaction::LoadedInteractive,
    runs_the_world: bool,
    can_activate: bool,
) -> GraphQLInteractive {
    let available = crate::interaction::is_available(loaded.row.effect_id.as_deref());
    let row = loaded.row;
    GraphQLInteractive {
        interactive_id: row.interactive_id,
        scene_id: row.scene_id,
        subject_kind: row.subject_kind,
        subject_ref: row.subject_ref,
        geometry: if runs_the_world {
            row.geometry.map(Json)
        } else {
            None
        },
        trigger: row.trigger,
        effect_id: if runs_the_world { row.effect_id } else { None },
        effect_config: if runs_the_world {
            row.effect_config.map(Json)
        } else {
            None
        },
        activation: runs_the_world.then_some(row.activation),
        fire_mode: runs_the_world.then_some(row.fire_mode),
        fired_at: if runs_the_world { row.fired_at } else { None },
        available: runs_the_world.then_some(available),
        can_activate,
    }
}

pub fn declaration_to_graphql(
    declaration: &thunderforge_canvas_core::interaction::EffectDeclaration,
) -> GraphQLEffectDeclaration {
    GraphQLEffectDeclaration {
        id: declaration.id.clone(),
        label: declaration.label.clone(),
        description: declaration.description.clone(),
        subject_kinds: declaration
            .subject_kinds
            .iter()
            .map(|k| k.as_str().to_string())
            .collect(),
        config: declaration
            .config
            .iter()
            .map(|field| {
                let (kind, reference_of, options) = match &field.kind {
                    ConfigFieldKind::Boolean => ("boolean", None, None),
                    ConfigFieldKind::Choice { options } => (
                        "choice",
                        None,
                        Some(
                            options
                                .iter()
                                .map(|o| GraphQLChoiceOption {
                                    value: o.value.clone(),
                                    label: o.label.clone(),
                                })
                                .collect(),
                        ),
                    ),
                    ConfigFieldKind::Reference { of } => ("reference", Some(of.clone()), None),
                    ConfigFieldKind::ReferenceList { of } => {
                        ("referenceList", Some(of.clone()), None)
                    }
                };
                GraphQLConfigField {
                    key: field.key.clone(),
                    label: field.label.clone(),
                    kind: kind.to_string(),
                    reference_of,
                    options,
                    required: field.required,
                }
            })
            .collect(),
    }
}
